weather_dl/download_pipeline/partition_test.py

```diff
@@ -0,0 +1,262 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import json
+import tempfile
+import typing as t
+import unittest
+from unittest.mock import MagicMock
+
+import apache_beam as beam
+from xarray_beam._src.test_util import EagerPipeline
+
+from .manifest import MockManifest, Location, LocalManifest
+from .parsers import get_subsections
+from .partition import skip_partition, PartitionConfig
+from .stores import InMemoryStore, Store
+
+
+class OddFilesDoNotExistStore(InMemoryStore):
+    def __init__(self):
+        super().__init__()
+        self.count = 0
+
+    def exists(self, filename: str) -> bool:
+        ret = self.count % 2 == 0
+        self.count += 1
+        return ret
+
+
+class PreparePartitionTest(unittest.TestCase):
+
+    def setUp(self) -> None:
+        self.dummy_manifest = MockManifest(Location('mock://dummy'))
+
+    def create_partition_configs(self, config, store: t.Optional[Store] = None) -> t.List[t.Dict]:
+        subsections = get_subsections(config)
+        params_cycle = itertools.cycle(subsections)
+
+        return (EagerPipeline()
+                | beam.Create([config])
+                | PartitionConfig(store, params_cycle, self.dummy_manifest))
+
+    def test_partition_single_key(self):
+        config = {
+            'parameters': {
+                'partition_keys': ['year'],
+                'target_path': 'download-{}.nc',
+            },
+            'selection': {
+                'features': ['pressure', 'temperature', 'wind_speed_U', 'wind_speed_V'],
+                'month': [str(i) for i in range(1, 13)],
+                'year': [str(i) for i in range(2015, 2021)]
+            }
+        }
+
+        actual = self.create_partition_configs(config)
+
+        self.assertListEqual([d['selection'] for d in actual], [
+            {**config['selection'], **{'year': [str(i)]}}
+            for i in range(2015, 2021)
+        ])
+
+    def test_partition_multi_key(self):
+        config = {
+            'parameters': {
+                'partition_keys': ['year', 'month'],
+                'target_path': 'download-{}-{}.nc',
+            },
+            'selection': {
+                'features': ['pressure', 'temperature', 'wind_speed_U', 'wind_speed_V'],
+                'month': [str(i) for i in range(1, 3)],
+                'year': [str(i) for i in range(2015, 2017)]
+            }
+        }
+
+        actual = self.create_partition_configs(config)
+
+        self.assertListEqual([d['selection'] for d in actual], [
+            {**config['selection'], **{'year': ['2015'], 'month': ['1']}},
```

Comment:
I know this is a more compact representation of the expected value, but I prefer the more explicit definition below.

Reply:
I think after we convert these to dataclasses, the representation will be much cleaner.